### ISSUE TITLE
Better testing idea

### ISSUE THREAD

**Issue by freesig** [FuelLabs/fuel-vm#277]
# Problem
Almost all functions in the vm take `self`. This makes unit test very difficult and results in poor test coverage.
# Idea
Only pass in what is actually needed.
From this: https://github.com/FuelLabs/fuel-vm/blob/c67688b097480f83bae33c54435b4ca9acbcd965/src/interpreter/alu.rs#L13
To this:
```rust
    /// Stores the overflowed wrapped value into REG_OF
    pub(crate) fn alu_capture_overflow<F, B, C>(
        &mut self,
        ra: RegisterId,
        f: F,
        b: B,
        c: C,
    ) -> Result<(), RuntimeError>
    where
        F: FnOnce(B, C) -> (u128, bool),
    {
        let is_wrapping = self.is_wrapping();
        alu_capture_overflow_inner(&mut self.registers, ra, is_wrapping, f, b, c)
    }
```
which calls into a pure function like:
```rust
/// Stores the overflowed wrapped value into REG_OF
pub(crate) fn alu_capture_overflow_inner<F, B, C, R>(
    registers: &mut R,
    ra: RegisterId,
    is_wrapping: bool,
    f: F,
    b: B,
    c: C,
) -> Result<(), RuntimeError>
where
    F: FnOnce(B, C) -> (u128, bool),
    R: Reg<REG_OF> + Reg<REG_ERR> + Reg<REG_PC>,
{
    is_register_writable(ra)?;
    let (result, _overflow) = f(b, c);

    if result > Word::MAX as u128 && !is_wrapping {
        return Err(PanicReason::ArithmeticOverflow.into());
    }

    // set the OF register to high bits of the u128 result
    *<R as Reg<REG_OF>>::get_mut(registers) = (result >> 64) as u64;
    *<R as Reg<REG_ERR>>::get_mut(registers) = 0;

    // set the return value to the low bits of the u128 result
    *<R as IndexMut<_>>::index_mut(registers, ra) = (result & Word::MAX as u128) as u64;

    registers.inc_pc()
}
```
This allows us to write tests like:
```rust
#[test_case(
    Input {
        ra: 0x13, is_wrapping: false, b: 1, c: 2,
        ..Default::default()
    } => matches Ok(Output{ pc: 4, overflow: 0, error: 0, result: 3 })
)]
#[test_case(
    Input {
        ra: 0x13, is_wrapping: false, b: u64::MAX, c: 2,
        ..Default::default()
    } => matches Err(RuntimeError::Recoverable(PanicReason::ArithmeticOverflow)) 
)]
#[test_case(
    Input {
        ra: 0x13, is_wrapping: true, b: u64::MAX, c: 2,
        ..Default::default()
    } => matches Ok(Output{ pc: 4, overflow: 2, error: 0, result: u64::MAX })
)]
fn test_alu_capture_overflow(mut input: Input) -> Result<Output, RuntimeError> {
    let ra = input.ra;
    let b = input.b;
    let c = input.c;
    let is_wrapping = input.is_wrapping;
    let f = u128::overflowing_add;

    alu_capture_overflow_inner(&mut input, ra, is_wrapping, f, b.into(), c.into())?;
    Ok(input.into())
}
```
_The last test case actually fails at the moment so maybe there's a bug?_

**Comment by freesig** [comment]
Types to understand the test code above.
```rust
#[derive(Default)]
struct Input {
    overflow: u64,
    error: u64,
    pc: u64,
    result: u64,
    ra: RegisterId,
    is_wrapping: bool,
    b: Word,
    c: Word,
}

#[derive(Default, Debug)]
struct Output {
    overflow: u64,
    error: u64,
    pc: u64,
    result: u64,
}
```

**Comment by freesig** [comment]
# Traits to encode register access
```rust

pub trait Reg<const I: usize>: IndexMut<usize, Output = u64> {
    fn get_mut(&mut self) -> &mut u64 {
        self.index_mut(I)
    }
}

impl Reg<REG_ERR> for [u64; VM_REGISTER_COUNT] {}
impl Reg<REG_PC> for [u64; VM_REGISTER_COUNT] {}
impl Reg<REG_OF> for [u64; VM_REGISTER_COUNT] {}

pub trait IncPc {
    fn inc_pc(&mut self) -> Result<(), RuntimeError>;
}

impl<T> IncPc for T
where
    T: Reg<REG_PC>,
{
    fn inc_pc(&mut self) -> Result<(), RuntimeError> {
        self.get_mut()
            .checked_add(fuel_asm::Instruction::LEN as Word)
            .ok_or_else(|| PanicReason::ArithmeticOverflow.into())
            .map(|pc| *self.get_mut() = pc)
    }
}
```

**Comment by freesig** [comment]
I think we could also get from:
```rust
*<R as Reg<REG_OF>>::get_mut(registers)
```
to 
```rust
*registers.get_err()
```

**Comment by freesig** [comment]
We could also fuzz / prop test these pure functions in a much easier way

**Comment by anton-trunov** [comment]
> We could also fuzz / prop test these pure functions in a much easier way

Totally agree! I think type-directed property-based randomized testing is a great tool to thoroughly test virtual machine and language implementations.

**Comment by Voxelot** [comment]
Traits for encoding register access seems very useful when mocking unit tests 💡

**Comment by freesig** [comment]
Yeh I figure we are not concerned with the vm compile times as it's such a small codebase so the more constraints we can move into the type system and therefore prove at compile time the better.

**Comment by freesig** [comment]
An interesting questing is where should we be checking the validity of the instructions. For example we know if the instruction is an `Add` then the `ra` register must be writable. Which is all things we actually know at the point we try and deserialize. You could even check these kind of constraints without de-serializing but I'm not sure how much value that would give.
What might be a good approach is to encode these types of invariants in the type of the `Instruction`. Then when we de-serialize it from memory we can exit early if any invariant is broken. This would mean the functions doing the work like `alu_capture_overflow` can't even be called with the wrong data (for example a `ra` register that is not writable). 
**TL/DR**
We constrain the function inputs as much as we can when de-serializing instructions to avoid those error cases.

**Comment by anton-trunov** [comment]
Incorporating constraints into types might be a nice use case for https://github.com/liquid-rust/flux

**Comment by freesig** [comment]
Yeh the problem with liquid types is it is a compile time check and we are mostly dealing with runtime data that is de-serialized into a type so there's no way to run the liquid type check after de-serialization.

**Comment by anton-trunov** [comment]
That's also something we can take care of. I think the best way to make sure serialization/deserialization is correct would be to use property-based randomized testing.

**Comment by Voxelot** [comment]
@freesig we should be able to close this issue now right? Seems like this has been mostly implemented now.
